### PR TITLE
clanchat: retrieve rank from ClanMember rather than ClanManager

### PR DESCRIPTION
**Pull request by trimbe**
Ranks would be cached at the time of join/leave which was incorrect. Using the rank from ClanMember will always be correct and won't require the user to have spoken recently to cache their rank.

### CODE REVIEW

**Comment by Adam-** [comment]
What was causing this to cache the old rank?

**Comment by trimbe** [comment]
Checking the rank in `ClanMemberJoined`/`ClanMemberLeft`.

File: runescape-api/src/main/java/net/runelite/rs/api/RSNameableContainer.java [hunk 1]
```diff
@@ -37,6 +37,9 @@
 	@Import("isMember")
 	boolean isMember(RSName var1);
 
+	@Import("findByName")
+	RSNameable findByName(RSName name);
```
Comment (Adam-) [inline runescape-api/src/main/java/net/runelite/rs/api/RSNameableContainer.java]:
```suggestion
	T findByName(RSName name);
```

File: runelite-mixins/src/main/java/net/runelite/mixins/RSClanMemberManagerMixin.java [hunk 2]
```diff
@@ -44,15 +45,27 @@
...
+		client.getCallbacks().postDeferred(event);
 	}
 
 	@Inject
 	@Override
 	public void rl$remove(RSNameable nameable)
 	{
-		ClanMemberLeft event = new ClanMemberLeft(nameable.getRsName().getName());
-		client.getCallbacks().post(event);
+		ClanMember member = (ClanMember) findByName(nameable.getRsName());
```
Comment (Adam-) [inline runelite-mixins/src/main/java/net/runelite/mixins/RSClanMemberManagerMixin.java]:
```suggestion
		ClanMember member = findByName(nameable.getRsName());
```

File: runelite-mixins/src/main/java/net/runelite/mixins/RSClanMemberManagerMixin.java [hunk 3]
```diff
@@ -44,15 +45,27 @@
 	@Override
 	public void rl$add(RSName name, RSName prevName)
 	{
-		ClanMemberJoined event = new ClanMemberJoined(name.getName());
-		client.getCallbacks().post(event);
+		ClanMember member = (ClanMember) findByName(name);
```
Comment (Adam-) [inline runelite-mixins/src/main/java/net/runelite/mixins/RSClanMemberManagerMixin.java]:
```suggestion
		ClanMember member = findByName(name);
```

File: runelite-api/src/main/java/net/runelite/api/events/ClanMemberJoined.java [hunk 4]
```diff
@@ -25,12 +25,13 @@
 package net.runelite.api.events;
 
 import lombok.Value;
+import net.runelite.api.ClanMember;
 
 @Value
 public class ClanMemberJoined
 {
 	/**
 	 * Name of the player who joined
```
Comment (Adam-) [inline runelite-api/src/main/java/net/runelite/api/events/ClanMemberJoined.java]: Update the javadoc

File: runelite-api/src/main/java/net/runelite/api/events/ClanMemberLeft.java [hunk 5]
```diff
@@ -25,12 +25,13 @@
 package net.runelite.api.events;
 
 import lombok.Value;
+import net.runelite.api.ClanMember;
 
 @Value
 public class ClanMemberLeft
 {
 	/**
 	 * Name of the player who left
```
Comment (Adam-) [inline runelite-api/src/main/java/net/runelite/api/events/ClanMemberLeft.java]: and here too

**Pull request merged**